Update docs (#254)
DigitecGalaxus/next-yak: packages/docs/content/docs/migration-from-styled-components.mdx

---
title: Migration from styled-components

> description: We tried to keep the changes you need to know if you're already familiar with styled-components to a minimum.

Reviewer: maybe we could also mention that this was one of the core goals of yak so we can switch from styled-components without losing all the know-how and dev tooling

For components with static styles, simply updating the import is sufficient:

```tsx title="component.tsx"
import styled from 'styled-components'; // [!code --]
import { styled } from 'next-yak'; // [!code ++]

const Button = styled.button`
  background: #BF4F74;
  color: white;
  padding: 1em 2em;
  border-radius: 4px;
  transition: all 0.2s ease-in-out;

  &:hover {
    transform: translateY(-1px);
    box-shadow: 0 2px 4px rgba(0,0,0,0.1);
  }
`;
```

<Callout>
**Important Note on Selectors**:
While most selectors work out of the box, you might encounter the error: `Selector "h1" is not pure (pure selectors must contain at least one local class or id)`. To resolve this, either:
1. Update your Next.js PostCSS version
2. Add an ampersand (&) before the selector

For more detailed information, refer to the [FAQ](/docs/faq#i-get-an-error-about-selector-that-are-not-pure).
</Callout>

For components with dynamic styles based on props:

```tsx title="component.tsx"
import { styled } from 'next-yak';

const Button = styled.button<{ $primary: boolean }>`
  background: #BF4F74;
  color: ${props => props.$primary ? "white" : "#BF4F74"};

  /* Alternative syntax */

>   ${props => props.$primary ? "color: white;" : "color: #BF4F74;"}

Reviewer: here ``css`...` `` is missing - which is also important when migrating - should we also mention this somewhere?

`;
```

See the [Dynamic component styles section](#dynamic-component-styles) for more details on handling dynamic styles.

### Static mixins

For static mixins, you can just change the import and you're done.

```tsx title="component.tsx"
import styled, { css } from 'styled-components'; // [!code --]
import { styled } from 'next-yak'; // [!code ++]

const mixin = css`
  color: green;
  font-size: 1rem;
`;

const MyComp = styled.div`
  background-color: yellow;
  ${mixin};
`;
```

This counts for all static mixins that you use in your styled components including exported mixins.

<Callout>
**Important note on semicolons**:
You need to add a semicolon after the mixin to separate it from the rest of the styles.

> Otherwise it will be treated as a part of the styles and our parser has no way to differentiate between the two.

Reviewer: ```suggestion
Otherwise the parser can not differentiate between a selector, a constant or a mixin.
```

</Callout>

> ### Attrs

Reviewer: Move after keyframes?

  // [!code highlight:1]
  animation: ${props => props.$rotate ? rotate : "none"} 2s linear infinite;
  padding: 2rem 1rem;
  font-size: 2rem;
`;
```

### Component references

Styling based on other components remains straightforward. You just need to change the import

```tsx title="component.tsx"
import styled from 'styled-components'; // [!code --]
import { styled } from 'next-yak'; // [!code ++]
import { Button } from './button';

const Container = styled.div`
  ${Button} {
    color: red;
    margin: 0.5em;

    &:hover {
      color: darkred;
    }
  }
`;
```

If you use the reference inside a dynamic component, you should have a look at the [Dynamic component styles section](#dynamic-component-styles) e.g.

```tsx title="component.tsx"
import styled from 'styled-components';
import { Button } from './button';

const DynamicContainer = styled.div<{ $variant: 'light' | 'dark' }>`
  // [!code highlight:13]
  ${props => props.$variant === 'light'
    ? css`
      ${Button} {
        color: #333;
        background: #fff;
      }
    `
    : css`
      ${Button} {
        color: #fff;
        background: #333;
      }
    `}
`;
```

## Move some code to yak files

> Other usages of styled-components need a bit more work. Yak files are special files that are evaluated during build time and can be used to store static values that are extracted during build time. You can create a `.yak.ts` file everywhere in your project and it will be picked up by next-yak.

Reviewer: ```suggestion
Migrating `styled-components` with complex logic might need a bit more work. Yak files are special files that are evaluated during build time and can be used to store static values that are extracted during build time. You can create a `.yak.ts` or `.yak.tsx` file everywhere in your project and it will be picked up by next-yak.

<Callout>
**Important note on `.yak.tsx` files**:
As yak files are evaluated on every code change, they will slow down your dev experience and your production build times. They should only be used as a last resort.
</Callout>
```

### Styles that use calculated values

Move calculated values to `.yak.ts` files:

```tsx title="component.tsx"
import styled from 'styled-components'; // [!code --]
import { styled } from 'next-yak'; // [!code ++]

const RADIUS = 5; // [!code --]
const CIRCUMFERENCE = 2 * Math.PI * RADIUS; // [!code --]
import { CIRCUMFERENCE } from './constants.yak'; // [!code ++]

const Circle = styled.div`
  width: ${CIRCUMFERENCE}px;
  height: ${CIRCUMFERENCE}px;
  border-radius: 50%;
  border: 1px solid black;
`;
```

```ts title="constants.yak.ts"
const RADIUS = 5; // [!code ++]
export const CIRCUMFERENCE = 2 * Math.PI * RADIUS; // [!code ++]
```

A general rule of thumb is that next-yak doesn't run any JavaScript code to figure out any interpolated values except the code recides in a `.yak.ts`(or `.yak.js`) file or is imported from a yak file.

### Shared values

Another use case for yak files is to store shared values that are used across your application.
Next-yak tries to optimize each file individually and doesn't share values between files.
So if you have a value that is used in multiple files, you should move it to a yak file so that it can be optimized across all files.

> 

Reviewer: We should explain that everything but calculations and string concatenations work WITHOUT yak files and should be preferred to keep the compilation fast

```tsx title="component.tsx"
import styled from 'styled-components'; // [!code --]
import { styled } from 'next-yak'; // [!code ++]
import { colors } from './constants'; // [!code --]
import { colors } from './constants.yak'; // [!code ++]

const Button = styled.button`
  background: ${colors.primary};
  color: white;
  padding: 1em 2em;
`;
```

```ts title="constants.yak.ts"
export const colors = { // [!code ++]
  primary: "#BF4F74", // [!code ++]
  secondary: "#F7B801", // [!code ++]
}; // [!code ++]
```

You can think of yak files as a special kind of file to optimize normal JavaScript values for styled components.

## Change some more things

The last part of the migration guide is about more complex cases where you need to change a bit more than just the import.

> ### CSS prop

Reviewer: ```suggestion
### CSS prop (optional)
```

If you use the css prop in your styled components, you can almost just change the import and you're done.
But you also need to tell TypeScript that native JSX elements can have a css prop. In order to do that, you can add the following to your `tsconfig.json`.

```json title="tsconfig.json"
{
  "compilerOptions": {
      "jsxImportSource": "next-yak" // [!code ++]
  }
}
```

Or if you want to use the css prop in a single file, you can just add the following to the top of the file.

```tsx title="component.tsx"
/** @jsxImportSource next-yak */ // [!code ++]
```

And afterwards you can just change the import and you're done.

```tsx title="component.tsx"
import styled from 'styled-components'; // [!code --]
import { css, styled } from 'next-yak'; // [!code ++]

const MyComponent = () => {
  return (
    <div
      css={css`
        background: papayawhip;
        color: red;
      `}
    />
  );
};
```

> Unfortunately, you can't use dynamic values in the css prop. If you need to use dynamic values, you should create a styled component with the styles from the css prop.

Reviewer: Instead show an example how dynamic runtime props work here

> ### Dynamic component styles

Reviewer: Maybe we should move this up (before yak files) under a Performance section

import styled from 'styled-components'; // [!code --]
import { styled } from 'next-yak'; // [!code ++]

const Button = styled.button<{ $primary: boolean }>`
  background: #BF4F74;
  color: ${props => props.$primary ? "white" : "#BF4F74"};
`;
```

This will be transformed to a css variable that gets injected during build time.

```tsx title="transformed.tsx"
// pseudo code
const Button = styled.button`
  background: #BF4F74;
  color: var(--next-yak-1);
`;
```

If you use dynamic css properties, you need to use the `css` tag to define the css.

```tsx title="component.tsx"
import styled, { css } from 'styled-components'; // [!code --]
import { styled } from 'next-yak'; // [!code ++]

const Button = styled.button<{ $primary: boolean }>`
  background: #BF4F74;
  ${props => props.$primary
  ? ` // [!code --]
  ? css` // [!code ++]
    color: white;
    font-size: 1rem;
    padding: 1em 2em;
  : ` // [!code --]
  : css` // [!code ++]
    color: #BF4F74;
    font-size: 2rem;
    padding: 2em 4em;
  `
  }
`;
```

When next-yak encounters a css tag, it will transform the css into a class name that gets injected during build time.

```tsx title="transformed.tsx"
// pseudo code

> const Button = (props) => props.$primary
>   ? <button className="next-yak-1">Click me</button>
>   : <button className="next-yak-2">Click me</button>
> ```

Reviewer: ```suggestion
const Button = (props) => props.$primary
  ? <button className="next-yak-1">Click me</button>
  : <button className="next-yak-2">Click me</button>
  
  
const Button = (props) => 
  <button className={props.$primary ? "next-yak-1" : "next-yak-2"}>
     Click me
  </button>
```
```

```tsx title="component.tsx"
import styled from 'styled-components'; // [!code --]
import { styled } from 'next-yak'; // [!code ++]

const Button = styled.button`
  color: blue;

  .myGlobalClass { // [!code --]
  :global(.myGlobalClass) { // [!code ++]
    color: red;
  }
`;
```

The reason for this is that next-yak transforms your styles to module-css styles. And if you want to reference global selectors, you need to use the `:global` selector.
More can be found in the [docs of css-modules](https://github.com/css-modules/css-modules/blob/master/docs/composition.md#exceptions)

### Functions that generate styles

If you have somewhere utility functions that generate styles, you need to change them. Either transform them to all available styles in a yak file or
create a dynamic mixin.

```tsx title="util.ts"
export const generateAllColors = (primaryColor: string) => { // [!code --]
  return { // [!code --]
    primary: primaryColor, // [!code --]
    secondary: "#F7B801", // [!code --]
  }; // [!code --]
}; // [!code --]
```

> ```tsx title="util.yak.ts"

Reviewer: I don't think this needs to be a `.yak` file

export const allColors = { // [!code ++]

>   "red": { // [!code ++]

Reviewer: I believe it's a constant in this case - so that should be fine
